sample_algorithms/preemptive.py

```diff
@@ -0,0 +1,270 @@
+import sys
+import time
+import math
+sys.path.append('../framework')
+
+import client
+import route as R
+import geojson
+import os.path
+
+from copy import deepcopy
+
+state = []
+locked_nodes = []
+
+SLEEP_TIME = 3
+TIMESLICE = 5
+MAX_SPEED_KM_H = 60
+INP_FILE = 'map.geojson'
+
+class ConnectionAssistant(client.SAVNConnectionAssistant):
+  def handleSimulationStart(self, initialParameters):
+    try: #TODO: Users should not need to try-except their code to get error messages
+      runSimulation(self, initialParameters)
+      #testInitialisation(initialParameters)
+    except Exception as err:
+      print(err)
+
+  def handleSimulationDataUpdate(self, update):
+    addToState(update['journeys'], state)
+
+  def handleSimulationCommunication(self, data):
+    analyseData(data)
+
+  def handleSimulationStop(self, info):
+    pass
+
+def postParams(initialParameters):
+  print(initialParameters)
+
+def testInitialisation(initialParameters):
+  global INP_FILE
+  n = 100
+  ts = 0
+
+  south = initialParameters['city']['bounds']['southWest']['lat']
+  west = initialParameters['city']['bounds']['southWest']['lng']
+  north = initialParameters['city']['bounds']['northEast']['lat']
+  east = initialParameters['city']['bounds']['northEast']['lng']
+  R.saveGeojson(south, west, north, east, INP_FILE)
+  INP_FILE = 'cpmap.geojson'
+  global state
+  for i in range(n):
+    os.system("cp map.geojson " + INP_FILE);
+    state = []
+    t = time.time()
+    addToState(initialParameters['journeys'], state)
+    t = time.time() - t
+    ts += t
+    print('Iteration ' + str(i) + ': ' + str(t) + ' seconds')
+  print('Average time: ' + str(ts / n))
+
+def runSimulation(savn, initialParameters):
+  south = initialParameters['city']['bounds']['southWest']['lat']
+  west = initialParameters['city']['bounds']['southWest']['lng']
+  north = initialParameters['city']['bounds']['northEast']['lat']
+  east = initialParameters['city']['bounds']['northEast']['lng']
+  print('Initialising geographical data')
+  R.saveGeojson(south, west, north, east, INP_FILE)
+  print('\t\t\t... Done')
+
+  print('Creating initial state')
+  global state
+  #state = initialParameters['state']
+
+  print('Preprocessing routes')
+  addToState(initialParameters['journeys'], state)
+  print('\t\t\t... Done')
+
+  print('Starting simulation:')
+  timestamp = initialParameters['timestamp']
+
+  print('\tSending data every ' + str(SLEEP_TIME) + ' seconds')
+
+  while savn.alive:
+    print("Sending", timestamp)
+    savn.updateState(timestamp, translate(state))
+    print("Working on next: Sent", timestamp)
+    state = executeGlobalAlgorithm(state)
+    timestamp += TIMESLICE
+  #useApiToEnd()
```

Comment:
you can call endSimulation from the parent SAVNConnectionAssistant to achieve a close me thinks